src/network/networkd-json.c

```diff
@@ -968,11 +907,11 @@ static int dhcp_server_append_json(Link *link, JsonVariant **v) {
         if (r < 0)
                 return r;
 
-        r = dhcp_server_offered_leases_append_json(link, &w);
+        r = dhcp_server_bound_leases_append_json(link->dhcp_server, &w);
```

Comment:
With a bunch of (nonsensical) leases I got a weird error message:
```sh
networkctl status 15 --json pretty 
Failed to get description: Structure needs cleaning
```
```sh
Got message type=method_call sender=:1.14972 destination=org.freedesktop.network1 path=/org/freedesktop/network1 interface=org.freedesktop.network1.Manager member=Describe  cookie=2 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Failed to build JSON data: Structure needs cleaning
Sent message type=error sender=n/a destination=:1.14972 path=n/a interface=n/a member=n/a cookie=56 reply_cookie=2 signature=s error-name=System.Error.EUCLEAN error-message=Structure needs cleaning
Failed to process message type=method_call sender=:1.14972 destination=org.freedesktop.network1 path=/org/freedesktop/network1 interface=org.freedesktop.network1.Manager member=Describe cookie=2 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a: Structure needs cleaning
```
`--json=off` seems to be fine:
```sh
networkctl status 15 --json off
...
         Offered DHCP leases: 192.168.18.156 (to ETHER)
                              192.168.18.152 (to ETHER)
                              192.168.18.158 (to ETHER)
                              192.168.18.150 (to ARP/LL)
                              192.168.18.148 (to ARP/LL)
                              192.168.18.151 (to ETHER)
                              192.168.18.153 (to ETHER)
                              192.168.18.146 (to ETHER)
                              192.168.18.157 (to ETHER)
                              192.168.18.154 (to ETHER)
                              192.168.18.149 (to ARP/LL)
                              192.168.18.155 (to ETHER)
                              192.168.102.171 (to IAID:0x5888c0c1/DUID)
                              192.168.18.147 (to ARP/LL)
```
When `networkd` was restarted 4 leases were restored:
```
         Offered DHCP leases: 192.168.18.149 (to ARP/LL)
                              192.168.18.147 (to ARP/LL)
                              192.168.18.155 (to ETHER)
                              192.168.102.171 (to IAID:0x5888c0c1/DUID)
```
and the `json` part worked too
```
networkctl status 15 --json pretty | jq .DHCPServer.Leases[].Address
[
  192,
  168,
  18,
  149
]
[
  192,
  168,
  18,
  147
]
[
  192,
  168,
  18,
  155
]
[
  192,
  168,
  102,
  171
]
```

Reply:
Looks like it's caused by non UTF-8 hostnames:
```python
>>> rdpcap('DHCP-non-utf-8-hostname')[0][DHCP].options
[('message-type', 1),
 ('client_id', b'\xffX\x88\xc0\xc1\x00\x02\x00\x00\xab\x11G\xa3]\xec\x905E9'),
 ('param_req_list', [1, 3, 6, 12, 15, 33, 42, 101, 114, 120, 121]),
 ('max_dhcp_size', 1472),
 ('rapid_commit', b''),
 ('hostname', b'\xff\xfe'),
 'end']
```
```
DHCPv4 server: DISCOVER (0x3c7950a8)
DHCPv4 server: ACK (0x3c7950a8)
DHCPv4 server: Failed to save leases, ignoring: Structure needs cleaning
```

Reply:
Thanks. Should be fixed now. Please test again.

Reply:
I'll take a look a bit later.

Having experimented with this PR I think it would be great to make it easier to clean up leases. Previously to get rid of nonsensical leases it was possible to just restart networkd. With this PR applied networkd has to be stopped and the`/var/lib/systemd/network/dhcp-server-lease/*` files have to be removed manually as far as I can tell.

Also based on my experiments networkd doesn't react to DHCP packets at all until the storage is ready. I'm not sure it's necessary when the DHCP server just relays packets. I'm not sure it should stop handing out leases with no storage either. My understanding is that the idea was to hand out leases (regardless of whether the storage is ready or not) and then save it once the "persistent storage" is ready. Dunno. I haven't fully read the thread yet.

Reply:
>  I'm not sure it's necessary when the DHCP server just relays packets.

Good point! Yeah, it is not necessary to wait for the persistent storage when the server is running in relay mode.

Reply:
> I'm not sure it should stop handing out leases with no storage either. My understanding is that the idea was to hand out leases (regardless of whether the storage is ready or not) and then save it once the "persistent storage" is ready. Dunno. I haven't fully read the thread yet.

If /var/lib/systemd/network is on a non-root partition, then we need to wait the partition is mounted. Otherwise, on reboot, the DHCP server may start before the partition is mounted, hence we cannot load previously bound leases, and an address already assigned to a host may be bound to another host.
So, we need to read the lease file before starting the DHCP server, and to support that we need to wait for the persistent storage being ready.

Reply:
> Having experimented with this PR I think it would be great to make it easier to clean up leases. Previously to get rid of nonsensical leases it was possible to just restart networkd. With this PR applied networkd has to be stopped and the`/var/lib/systemd/network/dhcp-server-lease/*` files have to be removed manually as far as I can tell.

Right. So, I also updated the our test script, to remove the persistent storage before starting networkd cleanly.

Reply:
Nonsensical leases aside I don't think it should be necessary to manually remove leases when containers are restarted for example. With (default) nspawn containers getting their IPs from networkd on the host `networkctl` looks really weird after those containers get rebooted. Below `192.168.46.129` gets mixed up with `192.168.249.177`.

The first run looks fine:
```
                     Address: 192.168.249.177
                              169.254.20.255
                              fe80::741a:afff:fe88:1910
           Activation Policy: up
         Required For Online: yes
                Connected To: C on port host0
         Offered DHCP leases: 192.168.249.182 (to ARP/LL)
                              192.168.249.183 (to ETHER)
                              192.168.249.180 (to ARP/LL)
                              192.168.249.187 (to ETHER)
                              192.168.249.179 (to ARP/LL)
                              192.168.249.185 (to ETHER)
                              192.168.249.178 (to ETHER)
                              192.168.249.189 (to ETHER)
                              192.168.249.184 (to IAID:0x5888c0c1/DUID)
                              192.168.249.188 (to ETHER)
                              192.168.249.186 (to ETHER)
                              192.168.249.181 (to ARP/LL)
                              192.168.249.190 (to ETHER)
```
The second run after the container was rebooted looks weird:
```
                     Address: 192.168.46.129
                              fe80::741a:afff:fe88:1910
           Activation Policy: up
         Required For Online: yes
                Connected To: C on port host0
         Offered DHCP leases: 192.168.249.178 (to ETHER)
                              192.168.249.181 (to ARP/LL)
                              192.168.249.189 (to ETHER)
                              192.168.249.187 (to ETHER)
                              192.168.249.180 (to ARP/LL)
                              192.168.249.182 (to ARP/LL)
                              192.168.249.188 (to ETHER)
                              192.168.249.186 (to ETHER)
                              192.168.249.185 (to ETHER)
                              192.168.249.183 (to ETHER)
                              192.168.249.179 (to ARP/LL)
                              192.168.46.136 (to IAID:0x5888c0c1/DUID)
                              192.168.249.190 (to ETHER)
```

Reply:
The "Structure needs cleaning" issue is gone as far as I can see.

Reply:
Interesting. But that's intended. Any leases should be dropped only when it is expired.
BTW, when picking up an IPv4 address requested with a null address, e.g. 0.0.0.0/24, I think we should take into account the lease file.